docs/standard/native-interop/tutorial-comwrappers.md

```diff
@@ -0,0 +1,430 @@
+        throw new NotImplementedException();
+}
+```
+
+To construct an instance of this class and provide it as a wrapper, you must define some policy. If this type is used as a wrapper, it would seem that since it implements both interfaces, the underlying COM instance should implement both interfaces too. Given that you're adopting this policy, you'll need to confirm this through calls to [`Marshal.QueryInterface()`][api_marshalqueryinterface] on the COM instance.
+
+```csharp
+int hr = Marshal.QueryInterface(ptr, ref IDemoGetType.IID_IDemoGetType, out IntPtr IDemoGetTypeInst);
+if (hr != 0)
+{
+    return null;
+}
+
+hr = Marshal.QueryInterface(ptr, ref IDemoStoreType.IID_IDemoStoreType, out IntPtr IDemoStoreTypeInst);
+if (hr != 0)
+{
+    Marshal.Release(ptr);
+    return null;
+}
+
+return new DemoNativeStaticWrapper()
+{
+    IDemoGetTypeInst = IDemoGetTypeInst,
+    IDemoStoreTypeInst = IDemoStoreTypeInst
+};
+```
+
+#### Dynamic Native Object Wrapper
+
+Dynamic wrappers are more flexible because they provide a way for types to be queried at run time instead of statically. In order to provide this support, you'll utilize [`IDynamicInterfaceCastable`][api_idynamicinterfacecastable]. Observe that `DemoNativeDynamicWrapper` only implements this interface. The functionality that the interface provides is a chance to determine what type is supported at run time.
+
+```csharp
+internal class DemoNativeDynamicWrapper
+    : IDynamicInterfaceCastable
+{
+    public RuntimeTypeHandle GetInterfaceImplementation(RuntimeTypeHandle interfaceType) =>
+        throw new NotImplementedException();
+
+    public bool IsInterfaceImplemented(RuntimeTypeHandle interfaceType, bool throwIfNotImplemented) =>
+        throw new NotImplementedException();
+}
+```
+
+Let's look at one of the interfaces that `DemoNativeDynamicWrapper` will dynamically support. The following code provides the implementation of `IDemoStoreType` using the *default interface methods* feature.
+
+```csharp
+[DynamicInterfaceCastableImplementation]
+unsafe interface IDemoStoreTypeNativeWrapper : IDemoStoreType
+{
+    public static void StoreString(IntPtr inst, int len, string? str);
+
+    void IDemoStoreType.StoreString(int len, string? str)
+    {
+        var inst = ((DemoNativeDynamicWrapper)this).IDemoStoreTypeInst;
+        StoreString(inst, len, str);
+    }
+}
+```
+
+There are two important things to note in this example:
+
+1) The `DynamicInterfaceCastableImplementationAttribute` attribute. This attribute is required on any type that is returned from a `IDynamicInterfaceCastable` method. It has the added benefit of making IL trimming easier, which means NativeAOT scenarios are more reliable.
+2) The cast to `DemoNativeDynamicWrapper`. This is part of the dynamic nature of `IDynamicInterfaceCastable`. The type that's returned from `IDynamicInterfaceCastable.GetInterfaceImplementation()` is used to blanket the type that implements `IDynamicInterfaceCastable`.
```

Comment:
> The type that's returned from `IDynamicInterfaceCastable.GetInterfaceImplementation()` is used to blanket the type that implements `IDynamicInterfaceCastable`.

It could be my subpar english, but I don't understand what this is trying to say. I know this is about the dynamic castable feature, which is not really com wrappers, so maybe this should link to another doc which explains how this works (I'm curious how this thing works).

Reply:
Not at all. It is a hard concept and my phrasing isn't great. I tried to explain it in a single sentence and likely didn't do a great job. I will try again.

I think we should have another tutorial on IDIC in general. /cc @elinor-fung

Reply:
Yeah, I'd agree that `IDynamicInterfaceCastable` is complicated and we should have a dedicated tutorial. (All we have is a [sample](https://github.com/dotnet/samples/tree/main/core/interop/IDynamicInterfaceCastable) and [part of a blog post](https://devblogs.microsoft.com/dotnet/improvements-in-native-code-interop-in-net-5-0/#idynamicinterfacecastable) from 5.0)